packages/field-base/src/helper-controller.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { SlotController } from '@vaadin/component-base/src/slot-controller.js';
+
+/**
+ * A controller to manage the helper node content.
+ */
+export class HelperController extends SlotController {
+  constructor(host) {
+    // Do not provide slot factory, as only create helper lazily.
+    super(host, 'helper');
+  }
+
+  get helperId() {
+    return this.node && this.node.id;
+  }
+
+  /**
+   * Override to initialize the newly added custom helper.
+   *
+   * @param {Node} helperNode
+   * @protected
+   * @override
+   */
+  initCustomNode(helperNode) {
+    this.__updateHelperId(helperNode);
+
+    this.__observeHelper(helperNode);
+
+    const hasHelper = this.__hasHelper(helperNode);
+    this.__toggleHasHelper(hasHelper);
+  }
+
+  /**
+   * Override to cleanup helper node when it's removed.
+   *
+   * @param {Node} _node
+   * @protected
+   * @override
+   */
+  teardownNode(_node) {
+    // The observer does not exist when the default helper is removed.
+    if (this.__helperIdObserver) {
+      this.__helperIdObserver.disconnect();
+    }
+
+    this.__applyDefaultHelper(this.helperText);
+  }
+
+  /**
+   * Set helper text based on corresponding host property.
+   * @param {string} helperText
+   */
+  setHelperText(helperText) {
+    this.helperText = helperText;
+
+    if (this.node === this.defaultNode) {
+      this.__applyDefaultHelper(helperText);
+    }
+  }
+
+  /**
+   * @param {HTMLElement} helperNode
+   * @return {boolean}
+   * @private
+   */
+  __hasHelper(helperNode) {
+    if (!helperNode) {
+      return false;
+    }
+
+    return helperNode.children.length > 0 || this.__isNotEmpty(helperNode.textContent);
+  }
+
+  /**
+   * @param {string} helperText
+   * @private
+   */
+  __isNotEmpty(helperText) {
+    return helperText && helperText.trim() !== '';
+  }
+
+  /**
+   * @param {string} helperText
+   * @private
+   */
+  __applyDefaultHelper(helperText) {
+    let helperNode = this.getSlotChild();
+
+    const hasHelperText = this.__isNotEmpty(helperText);
+
+    if (hasHelperText) {
+      if (!helperNode) {
+        // Set slot factory lazily to only create helper node when needed.
+        this.slotFactory = () => document.createElement('div');
+
+        helperNode = this.attachDefaultNode();
+
+        this.__updateHelperId(helperNode);
+        this.__observeHelper(helperNode);
+      }
+
+      if (helperNode === this.defaultNode) {
+        helperNode.textContent = helperText;
+      }
+    }
+
+    this.__toggleHasHelper(hasHelperText);
```

Comment:
I guess this should be only called when `this.node === this.defaultNode`.

Reply:
Updated the method so now it always uses default node.